### PR TITLE
[Movement] Collision: Path Planning Algorithm

### PR DESCRIPTION
**Pull request by maltewae**
Adds the implemtation for the path planning algorithm.

### CODE REVIEW

File: LabExT/Movement/PathPlanning.py [hunk 1]
```diff
@@ -284,3 +286,120 @@ def _find_lowest_potential(self) -> np.ndarray:
...
+                next_move[calibration] = potential_field.next_waypoint()
+
+            if self._last_moves_equal(
+                    self.last_moves[-abort_local_minimum:], next_move):
+                raise LocalMinimumError
+
+            self.last_moves.append(next_move)
+            yield next_move
+
+    def _get_grid_properties(self, tol: float = 0) -> tuple:
```
Comment (marcoep) [inline LabExT/Movement/PathPlanning.py:358]: call the `tol` parameter `padding`, as its not a tolerance but a "pad" around the maximum chip coordinates found.

File: LabExT/Movement/PathPlanning.py [hunk 2]
```diff
@@ -284,3 +286,120 @@ def _find_lowest_potential(self) -> np.ndarray:
...
+
+        Parameters
+        ----------
+        tol: float = 0
+            Grid outline tolerance.
+        """
+        all_points = np.concatenate([
+            [d._in_position for d in self.chip._devices.values()],
+            [d._out_position for d in self.chip._devices.values()]
+        ], axis=0)
```
Comment (marcoep) [inline LabExT/Movement/PathPlanning.py:373]: there is something off here, when I run this code, `all_points` is a 1D array? Then the two lines below for getting x and y coordinates do not make sense?

**Review by marcoep — CHANGES_REQUESTED** [review]

File: LabExT/Movement/PathPlanning.py (hunk 2 above)
Comment (maltewae) [reply LabExT/Movement/PathPlanning.py:437]: Really? I get a 2-dim array back. 🤔

File: LabExT/Movement/PathPlanning.py [hunk 3]
```diff
@@ -284,3 +286,120 @@ def _find_lowest_potential(self) -> np.ndarray:
...
+
+        xs = all_points[:, 0]
+        ys = all_points[:, 1]
+
+        outline = (
+            (xs.min() - tol, xs.max() + tol),  # X-min, X-max
+            (ys.min() - tol, ys.max() + tol)  # Y-min, Y-max
+        )
+        grid_size = np.floor(np.min(pdist(all_points)))
```
[the remark this answers is not in the document]
Comment (maltewae) [reply LabExT/Movement/PathPlanning.py]: Done!

File: LabExT/Movement/PathPlanning.py (hunk 1 above)
Comment (maltewae) [reply LabExT/Movement/PathPlanning.py]: Good idea!

File: LabExT/Movement/PathPlanning.py [hunk 4]
```diff
@@ -284,3 +286,120 @@ def _find_lowest_potential(self) -> np.ndarray:
...
+            self,
+            calibration: Type["Calibration"],
+            target: Type[ChipCoordinate]) -> None:
+        """
+        Registers a target for the given calibration.
+
+        Parameters
+        ----------
+        calibration : Calibration
+            Instance of a calibration
```
[the remark this answers is not in the document]
Comment (maltewae) [reply LabExT/Movement/PathPlanning.py]: Fixed!

File: LabExT/Movement/PathPlanning.py [hunk 5]
```diff
@@ -284,3 +286,120 @@ def _find_lowest_potential(self) -> np.ndarray:
...
+        self.potential_fields[calibration] = PotentialField(
+            calibration,
+            target,
+            self.grid_size,
+            self.grid_outline)
+
+    def trajectory(self, abort_local_minimum=3):
+        """
+
+        """
```
[the remark this answers is not in the document]
Comment (maltewae) [reply LabExT/Movement/PathPlanning.py]: Oops! I added a description of the function.

File: LabExT/Movement/PathPlanning.py (hunk 2 above)
Comment (marcoep) [reply LabExT/Movement/PathPlanning.py:437]: Sorry, you are correct, I was wrong.

**Review by marcoep — APPROVED** [review]

**Review by marcoep — APPROVED** [review]

**Review by marcoep — APPROVED** [review]

**Pull request merged**